File: gang/week13/올바른 괄호.java
import java.util.Stack;

class Solution {
    boolean solution(String s) {
        Stack<Character> bracket = new Stack<>();

        for(char b : s.toCharArray()) {
            if(b == '(') {  // 괄호가 ( 이면 push
                bracket.push(b);
            }
            else {
                if(bracket.isEmpty()) { // 만약 괄호가 ) 인데 스택이 비워져 있으면 괄호가 사라지지 않으므로 false반환
                    return false;
                }
                else {
                    bracket.pop();  // 비워져 있지 않으면 스택애는 ( 괄호가 있기 때문에 사라지므로 pop 사용
                }
            }
        }

        if(bracket.isEmpty()) { // 스택에 전부 비워져 있으면 true, 아니면 false 반환
            return true;
        }
        else {
            return false;
        }
    }
}

File: yongkeun/13week/올바른괄호.js
function solution(s) {
  // (면 배열에 넣고, )면 배열에서 제거하는 방식을 반복
  // )가 들어왔는데 배열에 아무것도 없으면 false
  // 마지막에 stack.length가 0이 아니여도 false(stack에 (이 남아있는 경우 false)
  let stack = [];
  for (let i = 0; i < s.length; i++) {
    if (s[i] === "(") {
      stack.push(1);
    } else if (s[i] === ")") {
      if (stack.length > 0) {
        stack.pop();
      } else {
        return false;
      }
    }
  }
  return stack.length === 0 ? true : false;
}
